### ISSUE TITLE
Strange `ocamldep` cli behaviour.

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/ocaml#9299]
It seems that `ocamldep` processes its files by side effect of [its `anon`ymous argument parsing function](https://github.com/ocaml/ocaml/blob/d2d94ef537bf9d18dc540a8fa09824b752bd8bb1/driver/makedepend.ml#L641). This looks a bit broken since `anon` is interleaved with option parsing arguments. So you may not have parsed everything  that influences the way you process the file. 

One obvious occurence of this is the list of include directories on [this line](https://github.com/ocaml/ocaml/blob/d2d94ef537bf9d18dc540a8fa09824b752bd8bb1/driver/makedepend.ml#L411) which may not be up-to-date yet. It's a bit difficult to see the extent of the problem since a lot of things happen via global variables and side effects, but there are at least others like `-pp` and `-ppx` args.

Just one example that shows processing before all `-I` have been consumed and thus different outcomes.
```
> echo "let () = B.v" > a.ml 
> mkdir b
> echo "let v = 3" > b/b.ml 
> ocamldep a.ml -I b
a.cmo :
a.cmx :
> ocamldep -I b a.ml 
a.cmo : \
    b/b.cmo
a.cmx : \
    b/b.cmx
```

I propose to make a PR that avoids this. Basically this means first collecting the source files in a list interleaved with the `-map` arguments which seem to be specifically context sensitive and then process the list.

**Comment by xavierleroy** [comment]
All OCaml programs that use the Arg module to process command lines have the behavior you describe: options and anonymous arguments are processed in the order in which they occur on the command-line.  To me this makes sense, this is consistent, this is easy to understand, and putting options after anonymous arguments is just silly to begin with.  If the consensus is that it is not silly and must be supported, maybe we should change the way the Arg module works rather than tweak every single program that uses it?

**Comment by dbuenzli** [comment]
> To me this makes sense,

I don't think it makes any sense.

Most command line interfaces are *not* context sensitive w.r.t. to positional and optional arguments and it's generally better they are so for usability reasons. For example putting optional arguments after positional ones perfectly makes sense in interactive scenarios where you simply add options to refine previous invocations.

I don't think this has anything to do with the way the Arg module works. It's just that generally one should not start running a program's service *while* you are parsing the cli. Parse your cli to gather the information in a datastructure and then run your service over it.

If I really need to justify this further then please explain to me why the third invocation below does not error like the first one:

```
> ocamlc ocamlcommon.cma                     
File "_none_", line 1:
Error: Cannot find file ocamlcommon.cma
> ocamlc -I +compiler-libs ocamlcommon.cma 
> ocamlc ocamlcommon.cma -I +compiler-libs     
```

**Comment by damiendoligez** [comment]
I have to side with @dbuenzli on this one. The `Arg` module is only a low-level interface for parsing the command line, it doesn't mandate the style of user interface that gets implemented. We've already migrated from a sequential to a declarative style for the compiler, we should also do it for `ocamldep`.

**Comment by dbuenzli** [comment]
Encouraged by @damiendoligez's lateral move, I cherry-picked my fix in pr #9795.
